.github/workflows/release.yml

```diff
@@ -0,0 +1,73 @@
+name: Build package
+
+on:
+  push:
+  pull_request:
+  release:
+    types:
+      - published
+  workflow_dispatch:
+
+permissions:
+  contents: read
+
+jobs:
+  # Always build & lint package.
+  build-package:
+    name: Build & verify package
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - uses: hynek/build-and-inspect-python-package@v2
+
+  # Publish to Test PyPI on every commit on main.
+  release-test-pypi:
+    name: Publish in-dev package to test.pypi.org
+    if: |
+      github.repository_owner == 'python'
+      && github.event_name == 'push'
+      && github.ref == 'refs/heads/main'
+    runs-on: ubuntu-latest
+    needs: build-package
+
+    permissions:
+      id-token: write
+
+    steps:
+      - name: Download packages built by build-and-inspect-python-package
+        uses: actions/download-artifact@v4
+        with:
+          name: Packages
+          path: dist
+
+      - name: Publish to Test PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+
+  # Publish to PyPI on GitHub Releases.
+  release-pypi:
+    name: Publish to PyPI
+    # Only run for published releases.
+    if: |
+      github.repository_owner == 'python'
+      && github.event.action == 'published'
+    runs-on: ubuntu-latest
+    needs: build-package
```

Comment:
I strongly recommend adding an extra layer of protection — GitHub Environments. Among other things, setting a deployment target URL makes it show up nicely in a few places in the GitHub web UI:

```suggestion
    needs: build-package

    environment:
      name: pypi
      url: >-
        https://pypi.org/project/blurb/${{
          github.event.release.tag_name
        }}
```

You can optionally require a human to hit an approval button though this deployment environment's protection rules in the repo settings, making it the last line of defense.

*(@woodruffw and I are trying to get more tutorials to showcase always using this from the very beginning, especially the official ones)*

P.S. Also consider publishing to GH Releases with Sigstore-produced signatures as shown in my guide: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#signing-the-distribution-packages.